KAFKA-6733: Printing additional ConsumerRecord fields in DefaultMessageFormatter

badaiaqrandista opened the pull request:
Implementation of KIP-431 - Support of printing additional ConsumerRecord fields in DefaultMessageFormatter

https://cwiki.apache.org/confluence/display/KAFKA/KIP-431%3A+Support+of+printing+additional+ConsumerRecord+fields+in+DefaultMessageFormatter

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

badaiaqrandista committed:
KAFKA-6733: Support of printing additional ConsumerRecord fields in DefaultMessageFormatter (rebased)

badaiaqrandista committed:
added tag for partition and offset to make them clearer

badaiaqrandista committed:
added code and test for null.literal property

dajac commented on core/src/main/scala/kafka/tools/ConsoleConsumer.scala:
```diff
@@ -459,48 +466,32 @@ class DefaultMessageFormatter extends MessageFormatter {
-    newProps
+  var headersDeserializer: Option[Deserializer[_]] = None
+
+  override def init(props: Properties): Unit = {
```
`init(props: Properties)` has been deprecated. It would be great if we could keep using `configure(configs: Map[String, _])` as before. I think that we should also try to directly extract the values from the `Map` instead of using a `Properties`.

badaiaqrandista replied:
@dajac 

I have replaced `init` with `configure` and changed code to extract the values directly from `Map`. Please review again.

dajac commented on core/src/test/scala/kafka/tools/DefaultMessageFormatterTest.scala:
```diff
@@ -0,0 +1,235 @@
+      Array(
+        "print all possible fields with default delimiters",
+        consumerRecord(),
+        Map("print.key" -> "true",
```
nit: Move `print.key` to next line to remain consistent with the formatting of the other Maps.

badaiaqrandista replied:
@dajac 

I've re-indent these as well.

dajac reviewed (commented):
@badaiaqrandista Thanks for the PR. It looks good overall. I left few comments.

badaiaqrandista committed:
Changed "init(props:Properties)" method to "configure(configs:Map[String,_])" method

badaiaqrandista committed:
Fix style in the test class

badaiaqrandista committed:
Modify DefaultMessageFormatter to get configs from Map instead of Properties

badaiaqrandista committed:
remove unused props variable

badaiaqrandista commented:
@dajac Can you please re-review the changes I've made?

bbejeck commented:
Ok to test

badaiaqrandista committed:
Fix error thrown by ConsoleConsumerTest and CustomDeserializerTest

badaiaqrandista commented:
@bbejeck 

One question: As part of my PR, I add a new test for DefaultMessageFormatter class: core/src/test/scala/kafka/tools/DefaultMessageFormatterTest.scala

But I found that there is another unit test that include a test for DefaultMessageFormatter class:
https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala#L501

This was the cause of the system test failure yesterday. I changed the latter test to work with the updated DefaultMessageFormatter. But after thinking about it further, I should just delete the test for DefaultMessageFormatter in ConsoleConsumerTest.scala.

What do you think?

bbejeck commented:
> I should just delete the test for DefaultMessageFormatter in ConsoleConsumerTest.scala.

@badaiaqrandista

I'm not sure.  My vote would be to keep the test but move it over to the `DefaultMessageFormatterTest.scala` class.  

But I'm not familiar enough with this code to say for sure.  From a quick look at the old test, it's not clear to me why it failed.  I guess the `Partition number` gets printed by default now?

\cc @dajac

badaiaqrandista commented:
@bbejeck 

It failed because it print the partition number as a single integer after the value. I moved the partition to be before the key (if printed) and value, and also added prefix "Partition:" to differentiate it from "Offset:".

It's only printed if "print.partition=true".

I will keep the tests as is then. ConsoleConsumerTest is a unit test for the class while DefaultMessageFormatterTest is more of an integration test.

bbejeck commented:
@badaiaqrandista can you rebase this PR?   We have a little bit of time to get this into 2.7 before the code freeze on 10/21

dajac commented:
Sorry guys. I have completely forgot this one. I think that it is fine to keep both tests as suggested by @badaiaqrandista.

dajac reviewed (commented):
@badaiaqrandista Thanks for the PR. The code LGTM pending Jenkins and the rebase.

bbejeck reviewed (commented):
Thanks for the PR @badaiaqrandista this LGTM modulo rebasing and a clean build.

bbejeck commented:
retest this please

bbejeck commented:
ok to test

bbejeck commented:
Java 11 and Java 15 passed.

Java 8 failed with `org.apache.kafka.streams.integration.PurgeRepartitionTopicIntegrationTest.shouldRestoreState` known flaky test.

badaiaqrandista commented:
@bbejeck Do I need to do anything on this?

bbejeck commented:
>Do I need to do anything on this?

@badaiaqrandista, nope as soon as we can get a green build, I'll merge it.

bbejeck commented:
Merged #9099 into trunk

bbejeck commented:
Thanks for the contribution @badaiaqrandista!
